### PR TITLE
Replicate followup

### LINKED ISSUE #4223: Ability to Duplicate Pages
# Story

As a content editor/webmaster, I would like to be able to duplicate an entire page so that I don't have to recreate it from scratch to have a consistent look.

# Additional Info

Replicate and Replicate UI modules have been discussed as a potential way forward.

Some additional information and conversations have been captured https://github.com/uiowa/uiowa/issues/2848.

Some more thoughts on possible issues with replicate https://github.com/uiowa/uiowa/pull/4214#issuecomment-929709156

# Proposed solution

Use replicate, replicate_ui, [this patch](https://www.drupal.org/project/replicate/issues/3100763), and possible a custom event listener to provide this functionality.

If entity_clone is used in place of replicate, there is a [patch here](https://www.drupal.org/project/entity_clone/issues/3050027) that is aimed at getting paragraphs nested under blocks working (may also be useful in setting up any necessary event subscribers for replicate).

### PR DESCRIPTION
**Pull request by cory-skeers**
Resolves #4223 

Re-enables replicate functionality for pages, articles, and resolves WSOD issue with paragraph replicants.

# How to test

Specifically, there's a WSOD issue under certain circumstances.

# Steps to reproduce:

- Replicate a page which uses either sliders or accordions (blocks w/ nested paragraphs)
- Publish the replicant page
- Delete the replicant page
- Attempt to view the original page, and get a WSOD

If the updates here work as intended, the above should be able to be completed without resulting in a WSOD.

## Original testing instructions

Below is the original list of functionality to be tested, and what should be preserved along with solving the WSOD issue.

- Checkout branch and composer install
- Sync a site and uli
- From a node page or article, or admin/content, choose to "replicate" -- enter a new title
- Check that the page has been successfully replicated
- All content should be present on the new page
- New blocks should be created
- New paragraphs nested under blocks should be created
- Section components should have new UUIDs
- New copies of referenced nodes (article authors, for instance), media, taxonomy terms, etc. should not be created, but the references should still be there
- Check the revision overview tab--there should be a single revision, with a message that it was replicated from another node (with a link to that node).
- Only Page and Article content types should be replicable. For other content types, you should not have the option to replicate.
- If you manually enter "node/#/replicate" for a non-replicable page, you should not be allowed to replicate, and should see a message stating this
- Publisher and webmaster should have replicate permissions. Related config splits that needed updating can be checked from https://github.com/uiowa/uiowa/search?p=1&q=filename%3Auser.role.webmaster or https://github.com/uiowa/uiowa/search?q=filename%3Auser.role.publisher
- Sitenow V2 sites should not have the option to replicate

# Followup tasks to create/update issues when this is merged

 * Remove temporary workaround for replicator::constant error
 * Update v2->v3 transition steps to include updating uiowa_core.replicate_allowed in uiowa_core.settings (see update hook 9020)

### CODE REVIEW

File: docroot/profiles/custom/sitenow/sitenow.install [hunk 1]
```diff
@@ -3539,3 +3539,30 @@ function sitenow_update_9021() {
     ->set('uiowa_core.replicate_allowed', [])
     ->save();
 }
+
+/**
+ * Re-activate replicate functionality for v3 sites.
```
Comment (joewhitsitt) [inline docroot/profiles/custom/sitenow/sitenow.install:3544]: You could simplify this if it is the same.

```php
function sitenow_update_9022() {
  sitenow_update_9020();
}
```

**Comment by cory-skeers** [comment]
My general understanding here is that the original replicate process prior to our EventSubscriber is creating at least a partial connection from the original paragraph to the replicant node's block, so that when the replicant node is deleted, so are its blocks, and then the blocks' paragraphs, including the original's paragraphs.

I think it's only partial, because we aren't seeing doubled-up paragraphs on the replicant, and are still seeing the paragraphs on the original. The only part that appears to not work as intended is how usage is being tracked upon deletion.

File: docroot/profiles/custom/sitenow/sitenow.install (hunk 1 above)
Comment (cory-skeers) [reply docroot/profiles/custom/sitenow/sitenow.install:3544]: Good call! Assuming we won't have transitioned all v2 sites by the time this rolls through, it'll be exactly the same. And if we did transition before then...that's either great for us on the v2->v3 move or wow...this issue was a doozy.

**Comment by cory-skeers** [comment]
> I think it's only partial, because we aren't seeing doubled-up paragraphs on the replicant, and are still seeing the paragraphs on the original. The only part that appears to not work as intended is how usage is being tracked upon deletion.

Paragraph's parent entity is being updated to the newly created block on the replicant, but all target_ids are correct.

So we either need to prevent the paragraph from being added/saved to the newly created replicant block, or re-set the original parent as part of our eventSubscriber.

**Comment by cory-skeers** [comment]
Added a comment to the patch issue (https://www.drupal.org/project/replicate/issues/3100763#comment-14256939)

**Comment by cory-skeers** [comment]
Included an updated patch that creates duplicates of entity_reference_revisions as well in the original listener patch.
https://www.drupal.org/project/replicate/issues/3100763#comment-14257205

**Comment by cory-skeers** [comment]
Want to look into where other issues may be at with replicator, and if there might be current in-progress work on the entity references where something like this may be better suited, or if it'd be better to attach a listener after the block duplication.

**Comment by cory-skeers** [comment]
Got a better understanding of the previous patch's "Create a duplicate of the inline block. For now we cannot use the Replicator service. The method "cloneByEntityRevisionId" has to be added." comment, and instead patched it to get around the issue and use the original Replicator service.

https://www.drupal.org/project/replicate/issues/3100763#comment-14257379

Still a bit more to look into, as well as want to get an MR going, but I think we're headed in the right direction.

**Comment by cory-skeers** [comment]
Currently getting a "cannot load entity with NULL id" error when the replicant is attempted to be saved. The null id is to be expected and understood, but not why it's being thrown when attempting to save.

**Comment by joewhitsitt** [comment]
plan on looking at this after standup

**Review by joewhitsitt — APPROVED** [review]
I tested the follow-up scenario as well as the general functionality again. Almost lost track of all the replicants i created/destroyed/altered/re-replicated. Also tried to cause chaos by removing/modifying new/replicated accordions.

Everything so far has held up.

**Pull request merged**